hw/xbox/nv2a.c

```diff
@@ -5661,55 +5704,67 @@ static void pgraph_write(void *opaque, hwaddr addr,
                                uint64_t val, unsigned int size)
 {
     NV2AState *d = opaque;
+    PGRAPHState *pg = &d->pgraph;
 
     reg_log_write(NV_PGRAPH, addr, val);
 
-    qemu_mutex_lock(&d->pgraph.lock);
+    qemu_mutex_lock(&pg->lock);
 
     switch (addr) {
     case NV_PGRAPH_INTR:
-        d->pgraph.pending_interrupts &= ~val;
-        qemu_cond_broadcast(&d->pgraph.interrupt_cond);
+        pg->pending_interrupts &= ~val;
+        qemu_cond_broadcast(&pg->interrupt_cond);
         break;
     case NV_PGRAPH_INTR_EN:
-        d->pgraph.enabled_interrupts = val;
+        pg->enabled_interrupts = val;
         break;
     case NV_PGRAPH_CTX_CONTROL:
-        d->pgraph.channel_valid = (val & NV_PGRAPH_CTX_CONTROL_CHID);
+        pg->channel_valid = (val & NV_PGRAPH_CTX_CONTROL_CHID);
         break;
     case NV_PGRAPH_CTX_USER:
         pgraph_set_context_user(d, val);
         break;
     case NV_PGRAPH_INCREMENT:
         if (val & NV_PGRAPH_INCREMENT_READ_3D) {
-            SET_MASK(d->pgraph.regs[NV_PGRAPH_SURFACE],
+            SET_MASK(pg->regs[NV_PGRAPH_SURFACE],
                      NV_PGRAPH_SURFACE_READ_3D,
-                     (GET_MASK(d->pgraph.regs[NV_PGRAPH_SURFACE],
+                     (GET_MASK(pg->regs[NV_PGRAPH_SURFACE],
                               NV_PGRAPH_SURFACE_READ_3D)+1)
-                        % GET_MASK(d->pgraph.regs[NV_PGRAPH_SURFACE],
+                        % GET_MASK(pg->regs[NV_PGRAPH_SURFACE],
                                    NV_PGRAPH_SURFACE_MODULO_3D) );
-            qemu_cond_broadcast(&d->pgraph.flip_3d);
+            qemu_cond_broadcast(&pg->flip_3d);
         }
         break;
     case NV_PGRAPH_FIFO:
-        d->pgraph.fifo_access = GET_MASK(val, NV_PGRAPH_FIFO_ACCESS);
-        qemu_cond_broadcast(&d->pgraph.fifo_access_cond);
+        pg->fifo_access = GET_MASK(val, NV_PGRAPH_FIFO_ACCESS);
+        qemu_cond_broadcast(&pg->fifo_access_cond);
         break;
+    case NV_PGRAPH_RDI_DATA: {
+        unsigned int select = GET_MASK(pg->regs[NV_PGRAPH_RDI_INDEX],
+                                       NV_PGRAPH_RDI_INDEX_SELECT);
+        unsigned int address = GET_MASK(pg->regs[NV_PGRAPH_RDI_INDEX],
+                                        NV_PGRAPH_RDI_INDEX_ADDRESS);
+        pgraph_rdi_write(pg, select, address, val);
+        assert(address < NV_PGRAPH_RDI_INDEX_ADDRESS); /* FIXME: Overflow into select? */
```

Comment:
This probably doesn't work due to the shift, it sohuld probably be something like `< GET_MASK(NV_PGRAPH_RDI_INDEX_ADDRESS, NV_PGRAPH_RDI_INDEX_ADDRESS)` to extract the number of bits